### PR TITLE
Set `service.instance.id` to `k8s.pod.uid` if it is not already set

### LINKED ISSUE #612: OTel pipeline improvements
* Set `service.instance.id` to `k8s.pod.uid` if not set
* Drop unneeded traces using a [Filter Processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/filterprocessor): `http.route= /live|/healthy|/ready`
  * Make this default and configurable
* Explore including [otelcol.processor.resourcedetection](https://grafana.com/docs/alloy/latest/reference/components/otelcol.processor.resourcedetection/) to gather cloud and resource information
  * Will need smart defaults and make it configurable

### PR DESCRIPTION
**Pull request by petewall**
Fixes #612

### CODE REVIEW

File: charts/k8s-monitoring-v1/docs/examples/alloy-autoscaling-and-storage/metrics.alloy [hunk 1]
```diff
@@ -84,6 +84,7 @@ otelcol.processor.transform "default" {
     context = "resource"
     statements = [
       "set(attributes[\"k8s.cluster.name\"], \"alloy-autoscaling-and-storage-test\") where attributes[\"k8s.cluster.name\"] == nil",
+      "set(attributes[\"service.instance.id\"], \"k8s.pod.id\") where attributes[\"service.instance.id\"] == nil",
     ]
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/docs/examples/alloy-autoscaling-and-storage/metrics.alloy:88]:
```suggestion
      "set(attributes[\"service.instance.id\"], attributes[\"k8s.pod.id\"]) where attributes[\"service.instance.id\"] == nil",
```

File: charts/k8s-monitoring-v1/docs/examples/alloy-autoscaling-and-storage/metrics.alloy [hunk 2]
```diff
@@ -84,6 +84,7 @@ otelcol.processor.transform "default" {
     context = "resource"
     statements = [
       "set(attributes[\"k8s.cluster.name\"], \"alloy-autoscaling-and-storage-test\") where attributes[\"k8s.cluster.name\"] == nil",
+      "set(attributes[\"service.instance.id\"], \"k8s.pod.id\") where attributes[\"service.instance.id\"] == nil",
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/docs/examples/alloy-autoscaling-and-storage/metrics.alloy:87]:
```suggestion
      "set(attributes[\"service.instance.id\"], attributes[\"k8s.pod.id\"]) where attributes[\"service.instance.id\"] == nil",
```

File: charts/k8s-monitoring-v1/docs/examples/alloy-autoscaling-and-storage/metrics.alloy [hunk 3]
```diff
@@ -93,6 +94,7 @@ otelcol.processor.transform "default" {
       "set(attributes[\"namespace\"], attributes[\"k8s.namespace.name\"])",
       "set(attributes[\"loki.resource.labels\"], \"cluster, namespace, job, pod\")",
       "set(attributes[\"k8s.cluster.name\"], \"alloy-autoscaling-and-storage-test\") where attributes[\"k8s.cluster.name\"] == nil",
+      "set(attributes[\"service.instance.id\"], \"k8s.pod.id\") where attributes[\"service.instance.id\"] == nil",
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/docs/examples/alloy-autoscaling-and-storage/metrics.alloy:97]:
```suggestion
      "set(attributes[\"service.instance.id\"], attributes[\"k8s.pod.id\"]) where attributes[\"service.instance.id\"] == nil",
```

File: charts/k8s-monitoring-v1/docs/examples/alloy-autoscaling-and-storage/output.yaml [hunk 4]
```diff
@@ -216,6 +216,7 @@ data:
         context = "resource"
         statements = [
           "set(attributes[\"k8s.cluster.name\"], \"alloy-autoscaling-and-storage-test\") where attributes[\"k8s.cluster.name\"] == nil",
+          "set(attributes[\"service.instance.id\"], \"k8s.pod.id\") where attributes[\"service.instance.id\"] == nil",
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/docs/examples/alloy-autoscaling-and-storage/output.yaml:219]:
```suggestion
          "set(attributes[\"service.instance.id\"], attributes[\"k8s.pod.id\"]) where attributes[\"service.instance.id\"] == nil",
```

**Review by cyrille-leclerc — CHANGES_REQUESTED** [review]
I think there is 2 bugs, I suggested changes on a few statements but not all of them as there are so many lines that look the same:
- It's `k8s.pod.uid` rather than `k8s.pod.id`. See [K8s Resource PRocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sattributesprocessor/) and [specs](https://opentelemetry.io/docs/specs/semconv/attributes-registry/k8s/#deprecated-kubernetes-attributes)
- It's `set(attributes["service.instance.id"], attributes["k8s.pod.uid"])` instead of `set(attributes["service.instance.id"], "k8s.pod.uid")`

cc @zeitlinger

**Comment by petewall** [comment]
> I think there is 2 bugs, I suggested changes on a few statements but not all of them as there are so many lines that look the same:
> 
> * It's `k8s.pod.uid` rather than `k8s.pod.id`. See [K8s Resource PRocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sattributesprocessor/?rgh-link-date=2024-10-08T07%3A27%3A38Z) and [specs](https://opentelemetry.io/docs/specs/semconv/attributes-registry/k8s/#deprecated-kubernetes-attributes)
> * It's `set(attributes["service.instance.id"], attributes["k8s.pod.uid"])` instead of `set(attributes["service.instance.id"], "k8s.pod.uid")`
> 
> cc @zeitlinger

Thanks Cyrille! You caught my late-night PR bugs. I merged in the changes and re-running tests.

File: charts/k8s-monitoring-v1/docs/examples/application-observability/metrics.alloy [hunk 5]
```diff
@@ -116,12 +117,14 @@ otelcol.processor.transform "default" {
       "set(attributes[\"namespace\"], attributes[\"k8s.namespace.name\"])",
       "set(attributes[\"loki.resource.labels\"], \"cluster, namespace, job, pod\")",
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/docs/examples/application-observability/metrics.alloy:118]: No longer relevant if we send OTLP logs to the Grafana Cloud OTLP Endpoint and adopt the new Loki format for OTLP logs.
`loki.resource.labels` will be ignored or, worse, added as a meaningless resource attribute ie a metadata.
This comment applies to all the files where this config block is added.

File: charts/k8s-monitoring-v1/docs/examples/application-observability/metrics.alloy [hunk 6]
```diff
@@ -116,12 +117,14 @@ otelcol.processor.transform "default" {
       "set(attributes[\"namespace\"], attributes[\"k8s.namespace.name\"])",
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/docs/examples/application-observability/metrics.alloy:117]: Setting `pod` and `namespace` as log resource attributes is very likely to no longer relevant if we send OTLP logs to the Grafana Cloud OTLP Endpoint and adopt the new Loki format for OTLP logs.
`pod` and `namespace` will become log metadata which requires a different LogQL query.
This comment applies to all the files where this config block is added.

File: charts/k8s-monitoring-v1/docs/examples/bearer-token-auth/metrics.alloy [hunk 7]
```diff
@@ -93,6 +94,7 @@ otelcol.processor.transform "default" {
       "set(attributes[\"namespace\"], attributes[\"k8s.namespace.name\"])",
       "set(attributes[\"loki.resource.labels\"], \"cluster, namespace, job, pod\")",
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/docs/examples/bearer-token-auth/metrics.alloy:95]: ditto

File: charts/k8s-monitoring-v1/docs/examples/bearer-token-auth/metrics.alloy [hunk 8]
```diff
@@ -93,6 +94,7 @@ otelcol.processor.transform "default" {
       "set(attributes[\"namespace\"], attributes[\"k8s.namespace.name\"])",
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/docs/examples/bearer-token-auth/metrics.alloy:94]: ditto

File: charts/k8s-monitoring-v1/templates/alloy_config/_processors.alloy.txt [hunk 9]
```diff
@@ -173,6 +174,7 @@ otelcol.processor.transform "default" {
       "set(attributes[\"namespace\"], attributes[\"k8s.namespace.name\"])",
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/templates/alloy_config/_processors.alloy.txt:174]: Setting `pod` and `namespace` as log resource attributes is very likely to no longer relevant if we send OTLP logs to the Grafana Cloud OTLP Endpoint and adopt the new Loki format for OTLP logs.
`pod` and `namespace` will become log metadata which requires a different LogQL query.
This comment applies to all the files where this config block is added.

File: charts/k8s-monitoring-v1/templates/alloy_config/_processors.alloy.txt [hunk 10]
```diff
@@ -173,6 +174,7 @@ otelcol.processor.transform "default" {
...
       "set(attributes[\"loki.resource.labels\"], \"{{ .Values.logs.receiver.transforms.labels | join ", " }}\")",
```
Comment (cyrille-leclerc) [inline charts/k8s-monitoring-v1/templates/alloy_config/_processors.alloy.txt:175]: No longer relevant if we send OTLP logs to the Grafana Cloud OTLP Endpoint and adopt the new Loki format for OTLP logs.
`loki.resource.labels` will be ignored or, worse, added as a meaningless resource attribute ie a metadata.
This comment applies to all the files where this config block is added.

**Review by cyrille-leclerc — APPROVED** [review]
The `service.instance.id` <-> `k8s.pod.uid` improvement looks good to me.
I identified things that I think should be cleaned up due to the adoption of the Grafana Cloud OTLP Endpoint.

**Comment by petewall** [comment]
@cyrille-leclerc , I'm going to extract your comments into a future improvement so it doesn't hold up this PR and release.
We'd want to make sure to remove those when using the OTLP gateway, but include them if we're not using them. This should be improved in Helm chart 2.0 anyway.

**Pull request merged**